Added convenience "Plugfigurable" class

Purg opened the pull request:
As the title says. This was created as a response to input that it was inconvenient to have to constantly descent from the two mixin classes. The name of this class is up for alternative suggestions, but this drew an endearing laugh from more than a few people, so here we are.

Purg commented:
PTAL @waxlamp + whoever

waxlamp reviewed (commented):
This will indeed be useful.

I am concerned about the name of the class: I personally like it, but I fear it will be confusing to the general audience. I wonder if we could brainstorm some alternatives before we settle on `Plugfigurable`.

`Pluggable` brings along a kind of runtime searchability, while `Configurable` means that you can reconstitute a thing from static data. Together, this suggests the idea of "instantiable" to me, but that's not much better than "plugfigurable".

The old name of "algorithm" is clearly too specific for this (an algorithm is a sequence of actions to take, and this combo is not that) but it has the advantage of concreteness. So the question here is: what is a concrete concept that captures the idea of both discoverability and instantiability? I think the right term might be "component". Is that term available in the ecosystem? Does it connote something different from what `Plugfigurable` provides?

Purg commented:
Do: check

Purg commented:
>  Together, this suggests the idea of "instantiable" to me, but that's not much better than "plugfigurable".

The "instantiable" term also seems at odds with the use of abstract classes (literally not instantiable).

"Component" is not currently taken as a term within SMQTK's existing ecosystem. I'm OK with using it, if only because I have no better ideas, but fair warning that it might come with it's own baggage depending on the user (I recall its a term used in other packages for different things (e.g. Qt)). Given that, we will need to of course add some docs on what specifically a "SMQTK Component" is and means to differentiate it from other contexts.

I will have to spend some time today/this weekend to generate some additional naming alternatives for consideration.

My only shred of a vote for "Plugfigurable" ("Plugfig"?) is that it's unique to my knowledge and remember-able. I'm definitely biased. Who doesn't remember "Pineapple Apple Pen". Also had a intro to CS prof that had a memorable "Cup-Chalk" lesson. I'm usually preferential to pretty direct terminology, so when I see "Plugfigurable" I'm reminded exactly what it does / brings in.

as6520 commented:
What about `ConfigurablePlugin`?

waxlamp commented:
> What about `ConfigurablePlugin`?

I think that "configurable plugin" has similar problems as "component", namely that it implies stuff that may not be true. Paul brings up a good point, that documentation will likely play a big role here. So with a name like "Plugfigurable", which is indeed memorable and unusual, perhaps the documentation can fill in the rest that's missing: an explanation of how this is a convenience interface that combines two of the lower level useful mixins.

Mainly because I am having trouble coming up with a better name, I'm somewhat backing off my request for something different here. I'll wait for Paul to think about this over the weekend and then we can make some kind of decision next week. Is that ok with you @Purg?

Purg commented:
Sounds good. I'll let names stew over the weekend and likely add some docs Monday.

Purg commented:
Stewing process did not yield anything new and wonderful. An earlier discussion today that included @waxlamp concluded that we might as well just run with "Plugfigurable" for the moment. Regardless of the specific name we use, it will come down to documentation to provide a clear meaning to whatever term we use for the interface. The current going value for "Plugfigurable" is that it's memorable.

I will work on adding the aforementioned documentation later today.

Purg commented:
@waxlamp @as6520 I've added a little documentation for the added class. It's pretty simple right now. I'm open to recommendations if there is more or a deeper explanation that should be conveyed. I also updated the documented example to utilize the new convenience class.

Purg commented:
I want to bump this for attention: I'm open to fixing any issues here, but if there are none currently, I'd like to get this merged so we can forge ahead.

Purg commented:
Me forgetting to mark as "Ready" probably didn't help.

as6520 commented:
Should there be a test added for "plugfigurable"? Apart from that LGTM

Purg commented:
I did not add any unittests due to the class not adding any meaningful functionality to be tested. Maybe a sort of end-to-end test might be applicable here? I can if there is consensus beyond just me, however I believe they would be mostly in duplicate with the unittests that exist for the underlying interfaces, besides maybe bringing together and "all at once" context."

as6520 commented:
If there isn't anything meaningful to test then there isn't any need to add test, just wanted to confirm that. LGTM

waxlamp commented on docs/plugins_configuration.rst:
```diff
@@ -138,9 +153,9 @@ package:
+   from smqtk_core import Plugfigurable
 
-   class MyInterface(Configurable, Pluggable):
+   class MyInterface(Plugfigurable):  # Transitively inherits from Pluggable and Configurable this is a really long line
```
Can you move this comment to the line above and remove the self-snark.

Purg replied:
Oh wow, I didn't mean for that to get through. I was testing a setting in my IDE! :D Let me get rid of that...

Purg replied:
As in move the comment to right above the class definition like this?
```python
   # Transitively inherits from Pluggable and Configurable
   class MyInterface(Plugfigurable):
       ...
```

waxlamp replied:
That's what I meant but it should actually go wherever it's supposed to (but I'd suggest not to put it where it was, even without the "lengthener").

Purg replied:
I'm not sure where it "should actually go," so I made the comment into a doc-string for the example class.

waxlamp reviewed (changes requested):
Just one minor comment about a comment. Looks great overall.
